state-chain/runtime/src/chainflip/bitcoin_block_processor.rs

```diff
@@ -43,46 +41,95 @@ type TypesVaultDepositWitnessing = TypesFor<BitcoinVaultDepositWitnessing>;
 
 type TypesEgressWitnessing = TypesFor<BitcoinEgressWitnessing>;
 
+/// Returns one event per deposit witness. If multiple events share the same deposit witness:
+/// - keep only the `Witness` variant,
+fn dedup_events<T: Ord + Clone>(
+	events: Vec<(BlockNumber, BtcEvent<T>)>,
+) -> Vec<(BlockNumber, BtcEvent<T>)> {
+	let mut chosen: BTreeMap<T, (BlockNumber, BtcEvent<T>)> = BTreeMap::new();
+
+	for (block, event) in events {
+		let deposit = event.deposit_witness().clone();
+
+		// Only insert if no event exists yet, or if we're upgrading from PreWitness to Witness
+		if !chosen.contains_key(&deposit) ||
+			(matches!(chosen.get(&deposit), Some((_, BtcEvent::PreWitness(_)))) &&
+				matches!(event, BtcEvent::Witness(_)))
+		{
+			chosen.insert(deposit, (block, event));
+		}
+	}
+
+	chosen.into_values().collect()
+}
 impl Hook<HookTypeFor<TypesDepositChannelWitnessing, ExecuteHook>>
 	for TypesDepositChannelWitnessing
 {
-	fn run(&mut self, (block, input): (BlockNumber, BtcEvent<DepositWitness<Bitcoin>>)) {
-		match input {
-			BtcEvent::PreWitness(deposit) => {
-				let _ = BitcoinIngressEgress::process_channel_deposit_prewitness(deposit, block);
-			},
-			BtcEvent::Witness(deposit) => {
-				BitcoinIngressEgress::process_channel_deposit_full_witness(deposit, block);
-			},
+	fn run(
+		&mut self,
+		events: Vec<(BlockNumber, BtcEvent<DepositWitness<Bitcoin>>)>,
+	) -> Vec<(BlockNumber, BtcEvent<DepositWitness<Bitcoin>>)> {
+		let deduped_events = dedup_events(events);
+		for (block, event) in &deduped_events {
+			match event {
+				BtcEvent::PreWitness(deposit) => {
+					let _ = BitcoinIngressEgress::process_channel_deposit_prewitness(
+						deposit.clone(),
+						*block,
+					);
+				},
+				BtcEvent::Witness(deposit) => {
+					BitcoinIngressEgress::process_channel_deposit_full_witness(
+						deposit.clone(),
+						*block,
+					);
+				},
+			}
 		}
+		deduped_events
```

Comment:
We could deduplicate the logic here by creating a generic `DedupExecHook<Number, Blockdata, Deposit, Prewitness: Hook<>, Witness: Hook<>>`

Then we could even instantiate this hook for testing.